### PR TITLE
Changes batch 2

### PR DESCRIPTION
**Pull request by facontidavide**
## Decorators

Decorators have a different meta-model when compared with Control nodes, since they allow only a single node. Therefore I introduced the class DecoratorNode.

## Clang format

Pure magic...

## Condition variables to wake up control nodes.

I removed the sleep in a loop and used condition_variable instead, where notify_all is triggered by __SetStatus__.

## Cleanups

It is always recommended to reduce as much as possible the scope of variables.

As agreed I renamed ReturnStatus to NodeSttatus to keep it shorter and More consistent with other enum names.

### CODE REVIEW

**Comment by facontidavide** [comment]
ops! unit test failed. I will investigate

Most probably it is the change in the condition variable.

**Comment by facontidavide** [comment]
Fixed :)

**Comment by facontidavide** [comment]
ready for review and merge, I will not make any further change to this PR, pinky promise ;)

**Comment by miccol** [comment]
ahahaha I am quite busy today. I will try to have a look into that this weekend (if it is not sunny :-P)

File: gtest/gtest_tree.cpp [hunk 1]
```diff
@@ -328,379 +310,352 @@ struct ComplexParallelTest : testing::Test
...
-    BT::ReturnStatus state = root->Tick();
+    BT::NodeStatus state = root->tick();
 
-    state = root->Tick();
+    state = root->tick();
 
-    std::this_thread::sleep_for(std::chrono::seconds(5));
-    state = root->Tick();
-    state = root->Tick();
+    std::this_thread::sleep_for(std::chrono::milliseconds(500));
```
Comment (miccol) [inline gtest/gtest_tree.cpp:375]: is that done to save time during tests?

Comment (facontidavide) [reply gtest/gtest_tree.cpp:375]: Yes, the only purpose. All the sleeps have been slashed to 1/10th.

File: src/control_node.cpp [hunk 2]
```diff
@@ -23,51 +20,49 @@ BT::ControlNode::ControlNode(std::string name) : TreeNode::TreeNode(name)
...
 {
-//    // Checking if the child is not already present
-//    for (unsigned int i=0; i<children_nodes_.size(); i++)
-//    {
-//        if (children_nodes_[i] == child)
-//        {
-//            throw BehaviorTreeException("'" + child->Name() + "' is already a '" + get_name() + "' child.");
-//        }
-//    }
+    // Checking if the child is not already present
```
Comment (miccol) [inline src/control_node.cpp]: Still puzzled whether the check should be there or not. Especially with nodes with memory, someone could find useful to have repeated actions e.g. SequenceMem(Look Object, Approach Object, Look Object, Grasp Object)

**Comment by miccol** [comment]
What is bt_factory?

**Comment by facontidavide** [comment]
__bt_factory__ is the preliminary skeleton (work in progress) of the class that I need to be able to load Behavior Trees from file.

It is just in "preview mode" in this PR, to allow you to take a look to the API.
As a reference, [this is the way](https://github.com/Eurecat-Robotics/BehaviorTree.CPP/blob/master/bt_core/sample/EX3_factory.cpp) it is used in my project. 

Take into account that the new version I am working on should be even simpler and flexible than the original one [here](https://github.com/Eurecat-Robotics/BehaviorTree.CPP/blob/master/bt_core/include/behavior_tree/BehaviorTree.h#L55)

If it bother you, I can exclude it from this PR and add it later on in a different PR.

**Comment by miccol** [comment]
Looks great!

File: src/control_node.cpp (hunk 2 above)
Comment (facontidavide) [reply src/control_node.cpp]: I want to discuss this with you in the future. The problem is that we make easier to create circular dependencies if we are not careful. Reusing the same object might have hard to debug consequences, but this PR is not the right time to discuss this.

Your call then: if you want, we can comment this again.

Comment (miccol) [reply src/control_node.cpp]: I would say let's comment this back and we can discuss that later.

**Comment by miccol** [comment]
Ready to merge. Waiting for the control_node.cpp part to be re-commented

**Comment by facontidavide** [comment]
commented

**Comment by miccol** [comment]
looks great!
Thanks

**Pull request merged**